presto-main/src/main/java/io/prestosql/type/TypeUtils.java

```diff
@@ -151,13 +160,46 @@ public static void checkElementNotNull(boolean isNull, String errorMsg)
     public static String getDisplayLabel(Type type, Session session)
     {
         if (isOmitDateTimeTypePrecision(session)) {
-            if (type instanceof TimestampType && ((TimestampType) type).getPrecision() == TimestampType.DEFAULT_PRECISION) {
-                return StandardTypes.TIMESTAMP;
-            }
-            else if (type instanceof TimestampWithTimeZoneType && ((TimestampWithTimeZoneType) type).getPrecision() == TimestampWithTimeZoneType.DEFAULT_PRECISION) {
-                return StandardTypes.TIMESTAMP_WITH_TIME_ZONE;
-            }
+            return getDisplayLabelForLegacyClients(type);
+        }
+        return type.getDisplayName();
+    }
+
+    private static String getDisplayLabelForLegacyClients(Type type)
+    {
+        if (type instanceof TimestampType && ((TimestampType) type).getPrecision() == TimestampType.DEFAULT_PRECISION) {
+            return StandardTypes.TIMESTAMP;
+        }
+        if (type instanceof TimestampWithTimeZoneType && ((TimestampWithTimeZoneType) type).getPrecision() == TimestampWithTimeZoneType.DEFAULT_PRECISION) {
+            return StandardTypes.TIMESTAMP_WITH_TIME_ZONE;
         }
+        if (type instanceof ArrayType) {
+            return ARRAY + "(" + getDisplayLabelForLegacyClients(((ArrayType) type).getElementType()) + ")";
+        }
+        if (type instanceof MapType) {
+            return MAP + "(" + getDisplayLabelForLegacyClients(((MapType) type).getKeyType()) + ", " + getDisplayLabelForLegacyClients(((MapType) type).getValueType()) + ")";
+        }
+        if (type instanceof RowType) {
+            return getRowDisplayLabelForLegacyClients((RowType) type);
+        }
+
         return type.getDisplayName();
     }
+
+    private static String getRowDisplayLabelForLegacyClients(RowType type)
+    {
+        List<String> fields = type.getFields().stream()
+                .map(field -> {
+                    String typeDisplayName = getDisplayLabelForLegacyClients(field.getType());
+                    if (field.getName().isPresent()) {
+                        return field.getName().get() + ' ' + typeDisplayName;
```

Comment:
Field names may need to be quoted. Have you checked how they are rendered currently?

Reply:
Yes. I've copied and refactored code from RowType